database/engine/journalfile.c

```diff
@@ -524,6 +521,614 @@ static uint64_t iterate_transactions(struct rrdengine_instance *ctx, struct rrde
+        count++;
+    }
+    return j2_extent_base + count;
+}
+
+static int verify_journal_space(struct journal_v2_header *j2_header, void *data, uint32_t bytes)
+{
+    if ((unsigned long)(((uint8_t *) data - (uint8_t *)  j2_header->data) + bytes) > (j2_header->total_file_size - sizeof(struct journal_v2_block_trailer)))
+        return 1;
+
+    return 0;
+}
+
+void *journal_v2_write_metric_page(struct journal_v2_header *j2_header, void *data, struct jv2_metrics_info *metric_info, uint32_t pages_offset)
+{
+    struct journal_metric_list *metric = (void *) data;
+
+    if (verify_journal_space(j2_header, data, sizeof(*metric)))
+        return NULL;
+
+    uuid_copy(metric->uuid, *metric_info->uuid);
+    metric->entries = metric_info->number_of_pages;
+    metric->page_offset = pages_offset;
+    metric->delta_start = (uint32_t)(metric_info->first_time_t - (time_t)(j2_header->start_time_ut / USEC_PER_SEC));
+    metric->delta_end = (uint32_t)(metric_info->last_time_t - (time_t)(j2_header->start_time_ut / USEC_PER_SEC));
+
+    return ++metric;
+}
+
+void *journal_v2_write_data_page_header(struct journal_v2_header *j2_header __maybe_unused, void *data, struct jv2_metrics_info *metric_info, uint32_t uuid_offset)
+{
+    struct journal_page_header *data_page_header = (void *) data;
+    uLong crc;
+
+    uuid_copy(data_page_header->uuid, *metric_info->uuid);
+    data_page_header->entries = metric_info->number_of_pages;
+    data_page_header->uuid_offset = uuid_offset;        // data header OFFSET poings to METRIC in the directory
+    data_page_header->crc = JOURVAL_V2_MAGIC;
+    crc = crc32(0L, Z_NULL, 0);
+    crc = crc32(crc, (void *) data_page_header, sizeof(*data_page_header));
+    crc32set(data_page_header->checksum, crc);
+    return ++data_page_header;
+}
+
+void *journal_v2_write_data_page_trailer(struct journal_v2_header *j2_header __maybe_unused, void *data, void *page_header)
+{
+    struct journal_page_header *data_page_header = (void *) page_header;
+    struct journal_v2_block_trailer *journal_trailer = (void *) data;
+    uLong crc;
+
+    crc = crc32(0L, Z_NULL, 0);
+    crc = crc32(crc, (uint8_t *) page_header + sizeof(struct journal_page_header), data_page_header->entries * sizeof(struct journal_page_list));
+    crc32set(journal_trailer->checksum, crc);
+    return ++journal_trailer;
+}
+
+void *journal_v2_write_data_page(struct journal_v2_header *j2_header, void *data, struct jv2_page_info *page_info)
+{
+    struct journal_page_list *data_page = data;
+
+    if (verify_journal_space(j2_header, data, sizeof(*data_page)))
+        return NULL;
+
+    struct extent_io_data *ei = page_info->custom_data;
+
+    data_page->delta_start_s = (uint32_t) (page_info->start_time_t - (time_t) (j2_header->start_time_ut) / USEC_PER_SEC);
+    data_page->delta_end_s = (uint32_t) (page_info->end_time_t -  (time_t) (j2_header->start_time_ut) / USEC_PER_SEC);
+    data_page->extent_index = page_info->extent_index;
+
```

Comment:
```sh
/database/engine/journalfile.c: 879 in journal_v2_write_data_page()
873         struct extent_io_data *ei = page_info->custom_data;
874     
875         data_page->delta_start_s = (uint32_t) (page_info->start_time_t - (time_t) (j2_header->start_time_ut) / USEC_PER_SEC);
876         data_page->delta_end_s = (uint32_t) (page_info->end_time_t -  (time_t) (j2_header->start_time_ut) / USEC_PER_SEC);
877         data_page->extent_index = page_info->extent_index;
878     
>>>     CID 433486:  High impact quality  (Y2K38_SAFETY)
>>>     A "time_t" value is stored in an integer with too few bits to accommodate it.  The expression "page_info->update_every" is cast to "uint32_t".
879         data_page->update_every_s = page_info->update_every;
880         data_page->page_length = (uint16_t) (ei ? ei->page_length : page_info->page_length);
881         data_page->type = 0;
882     
883         return ++data_page;
884     }
```

This is a common warning.

Reply:
Again about update every. It cannot get that big.